### ISSUE TITLE
Lost messages due to heartbeat handling

### ISSUE THREAD

**Issue by macboy012** [jasonrbriggs/stomp.py#94]
I have run into an issue where infrequently messages are delivered by ActiveMQ, but never seen in the consumer/listener. I eventually tracked this down to `__read()` in transport.py and the way heartbeats are detected/handled. I've been running version 4.1.2, however, the code in question isn't any different in 4.1.9.

Here is an example, the only important detail is the end. Each line represents one chunk from `self.receive()` in the `__read()` function:
```
Headers...
Body...
\n
\x00\n
```
The `\n` on line 3 (part of the message body) comes in as a single byte and matches the condition `elif c == b'\x0a':` triggering an immediate return, dumping everything in `fastbuf` and effectively losing the message.

My current solution is the following, inside of `elif c == b'\x0a':`:
```python
fastbuf_tell = fastbuf.tell()
if fastbuf_tell == 0 or (fastbuf_tell == 1 and fastbuf.getvalue() == b'\x0a'):
    # heartbeat (special case)
    return [c, ]
```
I don't know if it is a valid solution in all cases, only that it worked under my configuration.

If you take a hypothetical queue service that only writes one byte at a time to the connection, this problem would happen consistently, just in reading the headers. In real-world conditions this was happening about once every 4 hours on a 30 message/second connection.

One additional detail, this connection was configured with no heart beats: `heart-beat:0,0`. Ideally the heartbeat handling code would not be active when heart beats are turned off. This was the last place I looked because I figured, we don't have heart beats on, so I won't be hitting the heart beat special case.

**Comment by scop** [comment]
I think a full fix for this isn't solely about examining what was received in one read from the server, but rather what the receive buffer holds. The buffer might have partial frames in it, and if it does, then no matter what is received from the server, it should not be treated as a heartbeat until the partial frame is complete.

I've filed PR #120 which I believe fixes this, and while at it, adds support for CRLF heartbeats (allowed in STOMP 1.2). The test case code in it is quite ugly but should do its job at least for now. I haven't tested this beyond running the test suite, so approach with care. Look at the last commit of the PR for the actual fix for this, the others are less interesting, supporting ones with lots of noise caused by indentation level changes.

**Comment by scop** [comment]
Updated, more tested version of the fix is now in #120. The flaw with the previous one was that it treated optional EOLs following the NUL at end of frame as heartbeats. This one doesn't, is quite similar to macboy012's fix, but this one additionally avoids misdetections in the (unlikely) case that the "outer" receive buffer contains data when a heartbeat arrives. This one also removes the CRLF change, maybe to be handled later some time.

**Comment by kianxineki** [comment]
stomp.py 4.1.11
```
WARNING:stomp.py:Heartbeat timeout: diff_receive=7.020703365000372, time=23181.822471108, lastrec=23174.801767743
WARNING:stomp.py:Unable to issue SHUT_RDWR on socket because of error '[Errno 9] Bad file descriptor'
17
DEBUG:stomp.py:Sending a heartbeat message at 23182.823864987
DEBUG:stomp.py:Sending frame ['\n']
ERROR:stomp.py:Error sending frame
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/stomp/transport.py", line 565, in send
    self.socket.sendall(encoded_frame)
OSError: [Errno 9] Bad file descriptor
DEBUG:stomp.py:Unable to send heartbeat, due to: [Errno 9] Bad file descriptor
```

I get this error when active trace heartbeat (5000, 5000) and launched a task

not recover the error again and that message appears in infinite loop


```python
import time
import stomp
import logging


logging.basicConfig(level=10)


class MyListener(stomp.ConnectionListener):
    def on_error(self, headers, message):
        print('received an error "%s"' % message)

    def on_message(self, headers, message):
        print('received a message "%s"' % message)
        for x in range(50):
            print(x)
            time.sleep(1)
        print('finish')


conn = stomp.Connection(heartbeats=(4000, 4000))
conn.set_listener('', MyListener())
conn.start()
conn.connect('guest', 'guest', wait=True)
conn.subscribe(destination='/queue/rocket1', id=1, ack='auto')
conn.send('/queue/rocket1', 'OOooOoOoOhhh')

#while conn.is_connected():
time.sleep(60)
conn.disconnect()
```
